File: benchmarks/csharp/binary_trees.cs
// Binary trees benchmark adapted from The Computer Language Benchmarks Game

using System;
using System.Threading;

class BinaryTrees
{
    class TreeNode
    {
        public TreeNode left, right;
        public int item;
        
        public TreeNode(int item)
        {
            this.item = item;
        }
        
        // Create a new tree with the given item value at the root
        public static TreeNode CreateTree(int item, int depth)
        {
            TreeNode node = new TreeNode(item);
            
            if (depth > 0)
            {
                node.left = CreateTree(2 * item - 1, depth - 1);
                node.right = CreateTree(2 * item, depth - 1);
            }
            
            return node;
        }
        
        // Check the tree and return a checksum
        public int Check()
        {
            if (left == null)
            {
                return item;
            }
            return item + left.Check() - right.Check();
        }
    }
    
    public static void Main(string[] args)
    {
        int minDepth = 4;
        int maxDepth = 12;
        
        int stretchDepth = maxDepth + 1;
        DateTime startTime = DateTime.Now;
        
        // Allocate and check a big tree
        TreeNode bigTree = TreeNode.CreateTree(0, stretchDepth);
        Console.WriteLine("stretch tree of depth {0} check: {1}", stretchDepth, bigTree.Check());
        
        // Allow the big tree to be garbage collected
        bigTree = null;
        
        // Allocate a long-lived binary tree
        TreeNode longLivedTree = TreeNode.CreateTree(0, maxDepth);
        
        // Check trees of increasing depth
        for (int depth = minDepth; depth <= maxDepth; depth += 2)
        {
            int iterations = 1 << (maxDepth - depth + minDepth);
            int result = 0;
            
            for (int i = 0; i < iterations; i++)
            {
                TreeNode a = TreeNode.CreateTree(i, depth);
                TreeNode b = TreeNode.CreateTree(-i, depth);
                result += a.Check() + b.Check();
            }
            
            Console.WriteLine("{0} trees of depth {1} check: {2}", iterations * 2, depth, result);
        }
        
        // Check the long-lived tree last
        Console.WriteLine("long lived tree of depth {0} check: {1}", maxDepth, longLivedTree.Check());
        
        TimeSpan elapsed = DateTime.Now - startTime;
        Console.WriteLine("Time taken: {0} ms", elapsed.TotalMilliseconds);
        
        // Get memory stats
        long memoryUsed = GC.GetTotalMemory(true) / 1024;
        Console.WriteLine("Memory used: {0} KB", memoryUsed);
    }
}

File: benchmarks/javascript/binary_trees.js
// Binary trees benchmark adapted from The Computer Language Benchmarks Game

class TreeNode {
    constructor(item) {
        this.left = null;
        this.right = null;
        this.item = item;
    }
    
    // Create a new tree with the given item value at the root
    static createTree(item, depth) {
        const node = new TreeNode(item);
        
        if (depth > 0) {
            node.left = TreeNode.createTree(2 * item - 1, depth - 1);
            node.right = TreeNode.createTree(2 * item, depth - 1);
        }
        
        return node;
    }
    
    // Check the tree and return a checksum
    check() {
        if (!this.left) {
            return this.item;
        }
        return this.item + this.left.check() - this.right.check();
    }
}

function main() {
    const minDepth = 4;
    const maxDepth = 12;
    
    const stretchDepth = maxDepth + 1;
    const startTime = Date.now();
    
    // Allocate and check a big tree
    const bigTree = TreeNode.createTree(0, stretchDepth);
    console.log(`stretch tree of depth ${stretchDepth} check: ${bigTree.check()}`);
    
    // Allow the big tree to be garbage collected
    bigTree = null;
    
    // Allocate a long-lived binary tree
    const longLivedTree = TreeNode.createTree(0, maxDepth);
    
    // Check trees of increasing depth
    for (let depth = minDepth; depth <= maxDepth; depth += 2) {
        const iterations = 1 << (maxDepth - depth + minDepth);
        let result = 0;
        
        for (let i = 0; i < iterations; i++) {
            const a = TreeNode.createTree(i, depth);
            const b = TreeNode.createTree(-i, depth);
            result += a.check() + b.check();
        }
        
        console.log(`${iterations * 2} trees of depth ${depth} check: ${result}`);
    }
    
    // Check the long-lived tree last
    console.log(`long lived tree of depth ${maxDepth} check: ${longLivedTree.check()}`);
    
    const elapsed = Date.now() - startTime;
    console.log(`Time taken: ${elapsed} ms`);
    
    // Get memory stats - approximate since Node.js doesn't provide precise memory usage
    const memoryUsageInMB = process.memoryUsage().heapUsed / 1024 / 1024;
    console.log(`Memory used: ${Math.round(memoryUsageInMB * 1024)} KB`);
}

main();